packages/radio-group/src/vaadin-radio-group.js

```diff
@@ -73,510 +88,383 @@ class RadioGroupElement extends ThemableMixin(DirMixin(PolymerElement)) {
-  _addListeners() {
-    this.addEventListener('keydown', (e) => {
-      // if e.target is vaadin-radio-group then assign to checkedRadioButton currently checked radio button
-      let checkedRadioButton;
-
-      if (e.target === this) {
-        checkedRadioButton = this._checkedButton;
-      } else {
-        checkedRadioButton = e.composedPath().find((element) => {
-          return element instanceof RadioButton;
-        });
-      }
-      const horizontalRTL = this.getAttribute('dir') === 'rtl' && this.theme !== 'vertical';
-
-      // LEFT, UP - select previous radio button
-      if (e.keyCode === 37 || e.keyCode === 38) {
-        e.preventDefault();
-        this._selectIncButton(horizontalRTL, checkedRadioButton);
-      }
+  /**
+   * @param {KeyboardEvent} event
+   * @override
+   * @protected
+   */
+  _onKeyDown(event) {
+    super._onKeyDown(event);
 
-      // RIGHT, DOWN - select next radio button
-      if (e.keyCode === 39 || e.keyCode === 40) {
-        e.preventDefault();
-        this._selectIncButton(!horizontalRTL, checkedRadioButton);
-      }
+    const radioButton = event.composedPath().find((element) => {
+      return element instanceof RadioButton;
     });
 
-    this.addEventListener('focusin', () => this._setFocused(this._containsFocus()));
+    const isHorizontalRTL = this.getAttribute('dir') === 'rtl' && this.theme !== 'vertical';
 
-    this.addEventListener('focusout', () => {
-      this.validate();
-      this._setFocused(false);
-    });
+    // LEFT, UP - select previous radio button
+    if (['ArrowLeft', 'ArrowUp'].includes(event.key)) {
+      event.preventDefault();
+      this.__selectIncRadioButton(isHorizontalRTL, radioButton);
+    }
+
+    // RIGHT, DOWN - select next radio button
+    if (['ArrowRight', 'ArrowDown'].includes(event.key)) {
+      event.preventDefault();
+      this.__selectIncRadioButton(!isHorizontalRTL, radioButton);
+    }
   }
 
   /**
-   * @param {boolean} next
-   * @param {!RadioButton} checkedRadioButton
-   * @protected
+   * @param {!boolean} next
+   * @param {!RadioButton} radioButton
+   * @private
    */
-  _selectIncButton(next, checkedRadioButton) {
+  __selectIncRadioButton(next, radioButton) {
+    const index = this.__radioButtons.indexOf(radioButton);
+
     if (next) {
-      this._selectNextButton(checkedRadioButton);
+      this.__selectNextRadioButton(index);
     } else {
-      this._selectPreviousButton(checkedRadioButton);
+      this.__selectPrevRadioButton(index);
     }
   }
 
   /**
-   * @param {!RadioButton} element
-   * @param {boolean=} setFocusRing
-   * @protected
+   * @param {!number} index
+   * @private
    */
-  _selectButton(element, setFocusRing) {
-    if (this._containsFocus()) {
-      element.focus();
-      if (setFocusRing) {
-        element.setAttribute('focus-ring', '');
-      }
+  __selectNextRadioButton(index) {
+    const nextIndex = (this.__radioButtons.length + index + 1) % this.__radioButtons.length;
```

Comment:
The previous approach with `nextElementSibling`/ `prevElementSibling` didn't work with the slotted helper, label, and error message.